src/dialogs.js

```diff
@@ -230,85 +113,160 @@ class ClapperResumeDialog extends Gtk.MessageDialog
+        name: 'Media Files',
+    });
+    filter.add_mime_type('video/*');
+    filter.add_mime_type('audio/*');
+    filter.add_mime_type('application/claps');
+    Misc.subsMimes.forEach(mime => filter.add_mime_type(mime));
+    filters.append(filter);
+
+    const fileDialog = new Gtk.FileDialog({modal: true});
+    fileDialog.set_filters(filters);
+    fileDialog.open_multiple(window, null, _handleOpenLocal.bind(window.application));
+}
+
+function _handleOpenLocal(fileDialog, res)
+{
+    try {
+        const files = fileDialog.open_multiple_finish(res);
+        const filesArray = [];
 
-        super._init({
-            transient_for: window,
-            destroy_with_parent: true,
-            modal: true,
-            program_name: Misc.appName,
-            comments: _('A GNOME media player powered by GStreamer'),
-            version: pkg.version,
-            authors: ['Rafał Dzięgiel'],
-            artists: ['Rafał Dzięgiel'],
-            /* TRANSLATORS: Put your name(s) here for credits or leave untranslated */
-            translator_credits: _('translator-credits'),
-            license_type: Gtk.License.GPL_3_0,
-            logo_icon_name: 'com.github.rafostar.Clapper',
-            website: 'https://rafostar.github.io/clapper',
-            system_information: osInfo,
-        });
+        let index = 0;
+        let file;
 
-        this.show();
+        while((file = files.get_item(index))) {
+            filesArray.push(file);
+            index++;
+        }
+
+        const isHandlesOpen = Boolean(
+            this.flags & Gio.ApplicationFlags.HANDLES_OPEN
+        );
+
+        /* Remote app does not handle open */
+        if(isHandlesOpen)
+            this.open(filesArray, "");
+        else
+            this._openFilesAsync(filesArray);
     }
-});
+    catch(e) {
+        if(!e.matches(Gtk.DialogError.quark(), Gtk.DialogError.DISMISSED))
+            throw e;
+    }
+}
+
+function showExportPlaylistDialog(window)
+{
+    const filters = new Gio.ListStore();
+    const filter = new Gtk.FileFilter({
+        name: 'Playlist Files',
+    });
+    filter.add_mime_type('application/claps');
+    filters.append(filter);
+
+    const fileDialog = new Gtk.FileDialog({modal: true});
+    fileDialog.set_filters(filters);
+    fileDialog.set_initial_name('playlist.claps');
+    fileDialog.save(window, null, _handleExportPlaylist.bind(window.child.player.playlistWidget));
+}
+
+function _handleExportPlaylist(fileDialog, res)
+{
+    try {
+        const file = fileDialog.save_finish(res);
+        const playlist = this.getPlaylist(true);
+
+        FileOps.saveFileSimplePromise(file, playlist.join('\n'))
+            .then(() => {
+                debug(`exported playlist to file: ${file.get_path()}`);
+            })
+            .catch(err => {
+                debug(err);
+            });
+    }
+    catch(e) {
+        if(!e.matches(Gtk.DialogError.quark(), Gtk.DialogError.DISMISSED))
+            throw e;
+    }
```

Comment:
Not a big fan of putting everything in try-catch blocks. Especially when some errors are quietly ignored. GTK is single threaded and all this dialog stuff (including errors) should happen on main app thread, so AFAIK we either save or dismiss dialog (not both at once since this is single thread) and we already catch errors from file saving promise. Could you please elaborate why this is needed? Same question for other places where you used try-catch.

Reply:
`_handleExportPlaylist` is called by the Gtk.FileDialog regardless of whether a file was selected or the dialog was dismissed. Trying to call Gtk.FileDialog.save_finish when the dialog was dismissed then leads to an error, hence the try/catch block to filter out the dismissed error. I suppose we could always just ignore the error entirely